sqlglot/optimizer/annotate_types.py

```diff
@@ -270,10 +271,10 @@ class TypeAnnotator(metaclass=_TypeAnnotator):
         exp.Max: lambda self, e: self._annotate_by_args(e, "this", "expressions"),
         exp.Min: lambda self, e: self._annotate_by_args(e, "this", "expressions"),
         exp.Null: lambda self, e: self._annotate_with_type(e, exp.DataType.Type.NULL),
+        exp.Nullif: lambda self, e: self._annotate_by_args(e, "this"),
```

Comment:
This isn't correct for every dialect, e.g. postgres seems to have some coercion rules going on:

```
georgesittas=# select pg_typeof(nullif(1, 1.0));
 pg_typeof
-----------
 numeric
(1 row)

georgesittas=# select pg_typeof(nullif(1.0, 1));
 pg_typeof
-----------
 numeric
(1 row)

georgesittas=# select pg_typeof(nullif(1, 1));
 pg_typeof
-----------
 integer
(1 row)
```

Reply:
Perhaps it's safer to do a `_maybe_coerce` on the two args of `Nullif` or something

Reply:
Cool, seems like that might be what postgres is doing